swift/src/Ice/Connection.swift

```diff
@@ -169,40 +157,176 @@ public protocol Connection: AnyObject, CustomStringConvertible {
+        self.underlying = nil
+        self.incoming = incoming
+        self.adapterName = adapterName
+        self.connectionId = connectionId
+    }
+}
+
 /// Provides access to the connection details of an IP connection
-public protocol IPConnectionInfo: ConnectionInfo {
+open class IPConnectionInfo: ConnectionInfo {
     /// The local address.
-    var localAddress: String { get set }
+    public let localAddress: String
+
     /// The local port.
-    var localPort: Int32 { get set }
+    public let localPort: Int32
+
     /// The remote address.
-    var remoteAddress: String { get set }
+    public let remoteAddress: String
+
     /// The remote port.
-    var remotePort: Int32 { get set }
+    public let remotePort: Int32
+
+    public init(
+        incoming: Bool, adapterName: String, connectionId: String, localAddress: String, localPort: Int32,
+        remoteAddress: String, remotePort: Int32
+    ) {
+        self.localAddress = localAddress
+        self.localPort = localPort
+        self.remoteAddress = remoteAddress
+        self.remotePort = remotePort
+        super.init(incoming: incoming, adapterName: adapterName, connectionId: connectionId)
+    }
 }
 
 /// Provides access to the connection details of a TCP connection
-public protocol TCPConnectionInfo: IPConnectionInfo {
+public final class TCPConnectionInfo: IPConnectionInfo {
     /// The connection buffer receive size.
-    var rcvSize: Int32 { get set }
+    public let rcvSize: Int32
+
     /// The connection buffer send size.
-    var sndSize: Int32 { get set }
+    public let sndSize: Int32
+
+    internal init(
+        incoming: Bool, adapterName: String, connectionId: String, localAddress: String, localPort: Int32,
+        remoteAddress: String, remotePort: Int32, rcvSize: Int32, sndSize: Int32
+    ) {
+        self.rcvSize = rcvSize
+        self.sndSize = sndSize
+        super.init(
+            incoming: incoming, adapterName: adapterName, connectionId: connectionId, localAddress: localAddress,
+            localPort: localPort, remoteAddress: remoteAddress, remotePort: remotePort)
+    }
+}
+
+/// Provides access to the connection details of an SSL connection
+public final class SSLConnectionInfo: ConnectionInfo {
+    /// The certificate chain.
+    public let peerCertificate: SecCertificate?
+
+    internal init(underlying: ConnectionInfo, peerCertificate: String) {
+        let beginPrefix = "-----BEGIN CERTIFICATE-----\n"
+        let endPrefix = "\n-----END CERTIFICATE-----\n"
+
+        var raw = peerCertificate
+        if raw.hasPrefix(beginPrefix) {
+            raw = String(raw.dropFirst(beginPrefix.count))
+            raw = String(raw.dropLast(endPrefix.count))
+        }
+
+        var peerCertificate: SecCertificate? = nil
+        if let data = NSData(base64Encoded: raw, options: .ignoreUnknownCharacters) {
+            if let cert = SecCertificateCreateWithData(kCFAllocatorDefault, data) {
+                peerCertificate = cert
```

Comment:
Can't we just assign `self.peerCertificate` directly? the `var peerCertificate:` shadows the parameter, which is kind of confusing.

Reply:
We can only assign `self.peerCertificate` once, hence this helper.

An alternative would be to move this decoding to the factory.

Reply:
Fixed